AK/Random.cpp

```diff
@@ -1,68 +1,95 @@
 /*
  * Copyright (c) 2021, the SerenityOS developers.
+ * Copyright (c) 2024, the Ladybird developers.
+ * Copyright (c) 2025, Colleirose <criticskate@pm.me>
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <AK/Assertions.h>
 #include <AK/Platform.h>
 #include <AK/Random.h>
 #include <AK/UFixedBigInt.h>
 #include <AK/UFixedBigIntDivision.h>
 
+#if defined(AK_OS_LINUX)
+#    include <sys/random.h>
+#endif
+
 #if defined(AK_OS_WINDOWS)
 #    include <AK/NumericLimits.h>
 #    include <AK/Windows.h>
-#    include <bcrypt.h>
-#    include <ntstatus.h>
 #endif
 
-namespace AK {
-
-// NOTE: This function is supposed to always give a random number. If possible it is of good quality, but it can fall
-//       back to rand() if it fails on some systems. For high speed you should probably use a different generator.
-//       See MathObject::random() from LibJS. Where cryptographic security is needed use LibCrypto/SecureRandom.h.
-void fill_with_random([[maybe_unused]] Bytes bytes)
+static inline ErrorOr<void> csprng(void* const buf, size_t size)
 {
+    // We shouldn't use OpenSSL's RAND_bytes function here, because we want to avoid adding dependencies to AK.
+    // Therefore, we will use the best platform-specific CSPRNG.
 #if defined(AK_OS_SERENITY) || defined(AK_OS_ANDROID) || defined(AK_OS_BSD_GENERIC) || defined(AK_OS_HAIKU) || AK_LIBC_GLIBC_PREREQ(2, 36)
-    arc4random_buf(bytes.data(), bytes.size());
-#elif defined(OSS_FUZZ)
-#else
-    auto fill_with_random_fallback = [&]() {
-        for (auto& byte : bytes)
-            byte = rand();
-    };
+    // MacOS and iOS both appear to support arc4random_buf, and AK/Platform.h defines AK_OS_BSD_GENERIC when building for them,
+    // so they will be covered by the compile target that uses arc4random_buf.
+    //
+    // Note that although [SecRandomCopyBytes](https://developer.apple.com/documentation/security/secrandomcopybytes(_:_:_:)) appears to be more commonly used on those platforms,
+    // it doesn't seem to be required.
+    arc4random_buf(buf, size);
+#elif defined(AK_OS_LINUX)
+    int ret;
+    while (size > 0u) {
+        // The possible errors that can be returned here are:
+        //
+        // EINTR, interrupted by a signal handler.
+        // EAGAIN, the requested entropy wasn't available. Only possible if the GRND_NONBLOCK flag is set, which it isn't.
+        // EINVAL, invalid flags provided, which there aren't.
+        // ENOSYS, the kernel doesn't implement the syscall. The syscall has been available since Linux 3.17, which was released in 2019.
+        // EFAULT, the address is outside the accessible space; the caller is responsible for providing a valid address.
+        //
+        // Therefore, the only somewhat plausible errors are EFAULT and EINTR.
+        // Nonetheless, we will check for the possibility of all the error codes out of an abundance of caution.
```

Comment:
That's a lot of text just to say "we're checking all error cases" - let's put this in the commit message if it's really relevant enough.